netZooPy/panda/panda.py

```diff
@@ -405,30 +414,121 @@ def panda_loop(self, correlation_matrix, motif_matrix, ppi_matrix):
         def t_function(x, y=None):
             '''T function.'''
             if y is None:
-                a_matrix = np.dot(x, x.T)
-                s = np.square(x).sum(axis=1)
-                a_matrix /= np.sqrt(s + s.reshape(-1, 1) - np.abs(a_matrix))
+                a_matrix = xp.dot(x, x.T)
+                s = xp.square(x).sum(axis=1)
+                a_matrix /= xp.sqrt(s + s.reshape(-1, 1) - xp.abs(a_matrix))
+            else:
+                a_matrix = xp.dot(x, y)
+                a_matrix /= xp.sqrt(xp.square(y).sum(axis=0) + xp.square(x).sum(axis=1).reshape(-1, 1) - xp.abs(a_matrix))
+            return a_matrix
+
+        def update_diagonal(diagonal_matrix, num, alpha, step):
+            '''Update diagonal.'''
+            xp.fill_diagonal(diagonal_matrix, xp.nan)
+            diagonal_std = xp.nanstd(diagonal_matrix, 1)
+            diagonal_fill = diagonal_std * num * math.exp(2 * alpha * step)
+            xp.fill_diagonal(diagonal_matrix, diagonal_fill)
+
+        def gt_function(x, y=None):
+            '''T function.''' 
+            if y is None:
+                # a_matrix = tf.tensordot(x, tf.transpose(x),axes=1)
+                a_matrix = tf.linalg.matmul(x, tf.transpose(x))#x,transpose_b=True)
+                s = xp.square(x).sum(axis=1)
+                a_matrix /= xp.sqrt(s + s.reshape(-1, 1) - xp.abs(a_matrix.numpy()))
+            else:
+                # a_matrix = tf.linalg.matmul(x, y,axes=1)
+                a_matrix = tf.linalg.matmul(x, y)
+                a_matrix /= xp.sqrt(xp.square(y).sum(axis=0) + xp.square(x).sum(axis=1).reshape(-1, 1) - xp.abs(a_matrix.numpy()))
+            return a_matrix.numpy()
+
+        panda_loop_time = time.time()
+        num_tfs, num_genes = motif_matrix.shape
+        step = 0
+        hamming = 1
+        alpha = 0.1
```

Comment:
I can't believe that alpha is hard-coded in pypanda, we would have been better off writing our own software than fixing this